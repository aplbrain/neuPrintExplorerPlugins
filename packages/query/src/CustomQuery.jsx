/*
 * Supports simple, custom neo4j query.
 */
import React from 'react';
import PropTypes from 'prop-types';
import randomColor from 'randomcolor';
import { withRouter } from 'react-router';

import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';

const styles = () => ({
  textField: {
    width: 300,
    margin: '0 0 1em 0'
  },
  button: {
    margin: 4,
    display: 'block'
  },
  formControl: {}
});

const pluginName = 'CustomQuery';

export class CustomQuery extends React.Component {
  static get queryName() {
    return 'Custom';
  }

  static get queryDescription() {
    return 'Enter custom Neo4j Cypher query';
  }

  constructor(props) {
    super(props);
    const initqsParams = {
      textValue: ''
    };
    const qsParams = props.actions.LoadQueryString(
      `Query:${this.constructor.queryName}`,
      initqsParams,
      props.urlQueryString
    );
    this.state = {
      qsParams
    };
  }

  processResults = (query, apiResponse) => {
    if (apiResponse.data) {
      const data = apiResponse.data.map(row =>
        row.map(item => (typeof item === 'object' ? JSON.stringify(item) : item))
      );
      return {
        columns: apiResponse.columns,
        data,
        debug: apiResponse.debug
      };
    }
    return {
      columns: [],
      data: [],
      debug: ''
    };
  };

  processRequest = () => {
    const { dataSet, actions, history } = this.props;
    const { qsParams } = this.state;
    const { textValue } = qsParams;

    const query = {
      dataSet,
      cypherQuery: textValue,
      visType: 'SimpleTable',
      plugin: pluginName,
      parameters: {},
      title: 'Custom query',
      menuColor: randomColor({ luminosity: 'light', hue: 'random' }),
      processResults: this.processResults
    };
    actions.submit(query);
    history.push({
      pathname: '/results',
      search: actions.getQueryString()
    });
    return query;
  };

  handleChange = event => {
    const { actions } = this.props;
    const { qsParams } = this.state;
    const oldParams = qsParams;
    oldParams.textValue = event.target.value;
    actions.setURLQs(actions.SaveQueryString(`Query:${this.constructor.queryName}`, oldParams));
    this.setState({
      qsParams: oldParams
    });
  };

  catchReturn = event => {
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      this.processRequest();
    }
  };

  render() {
    const { qsParams } = this.state;
    const { classes, isQuerying } = this.props;
    return (
      <FormControl className={classes.formControl}>
        <TextField
          label="Custom Cypher Query"
          multiline
          value={qsParams.textValue}
          rows={1}
          rowsMax={4}
          className={classes.textField}
          onChange={this.handleChange}
          onKeyDown={this.catchReturn}
        />
        <Button
          variant="contained"
          className={classes.button}
          onClick={this.processRequest}
          color="primary"
          disabled={isQuerying}
        >
          Submit
        </Button>
      </FormControl>
    );
  }
}

CustomQuery.propTypes = {
  urlQueryString: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired,
  dataSet: PropTypes.string.isRequired,
  history: PropTypes.object.isRequired,
  isQuerying: PropTypes.bool.isRequired
};

export default withStyles(styles)(withRouter(CustomQuery));
