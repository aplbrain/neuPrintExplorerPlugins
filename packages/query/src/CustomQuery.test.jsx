import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { CustomQuery } from './CustomQuery';

let wrapper;
let button;
let textField;
let submit;

const actions = {
  submit: jest.fn(),
  LoadQueryString: jest.fn((_, initqsParams) => initqsParams),
  setURLQs: jest.fn(),
  SaveQueryString: jest.fn(),
  skeletonAddandOpen: jest.fn(),
  neuroglancerAddandOpen: jest.fn(),
  getQueryString: jest.fn(),
  getQueryObject: jest.fn(() => ({})),
  setQueryString: jest.fn(),
  metaInfoError: jest.fn()
};

const styles = { textField: '', button: '', formControl: '' };

const component = (
  <CustomQuery
    dataSet="test"
    history={{ push: jest.fn() }}
    classes={styles}
    actions={actions}
    urlQueryString=""
    isQuerying={false}
  />
);

describe('custom query Plugin', () => {
  beforeAll(() => {
    wrapper = mount(component);
    button = wrapper.find('CustomQuery').find('Button');
    textField = wrapper.find('CustomQuery').find('TextField');
    submit = jest.spyOn(wrapper.find('CustomQuery').props().actions, 'submit');
  });
  beforeEach(() => {
    submit.mockReset();
  });
  it('has name and description', () => {
    expect(CustomQuery.queryName).toBeTruthy();
    expect(CustomQuery.queryDescription).toBeTruthy();
  });
  it('renders correctly', () => {
    const pluginView = renderer.create(component).toJSON();
    expect(pluginView).toMatchSnapshot();
  });
  describe('when user clicks submit', () => {
    it('should return a query object and submit', () => {
      submit = jest.spyOn(wrapper.find('CustomQuery').props().actions, 'submit');
      expect(button.props().onClick()).toEqual(
        expect.objectContaining({
          dataSet: 'test',
          cypherQuery: '',
          visType: 'SimpleTable',
          plugin: 'CustomQuery',
          parameters: {},
          title: 'Custom query',
          menuColor: expect.any(String),
          processResults: expect.any(Function)
        })
      );
      expect(submit).toHaveBeenCalledTimes(1);
    });

    it('should process returned results into data object', () => {
      const query = {
        dataSet: 'test',
        cypherQuery: 'test',
        visType: 'SimpleTable',
        plugin: 'CustomQuery',
        parameters: {},
        title: 'Custom query'
      };
      const apiResponse = { data: [[1, 2, 3], [4, 5, 6]], columns: ['a', 'b', 'c'], debug: 'test' };
      const processedResults = wrapper
        .find('CustomQuery')
        .instance()
        .processResults(query, apiResponse);
      expect(processedResults).toEqual(apiResponse);

      // if no data returned
      const processedResultsEmpty = wrapper
        .find('CustomQuery')
        .instance()
        .processResults(query, {});
      expect(processedResultsEmpty).toEqual({
        columns: [],
        data: [],
        debug: ''
      });
    });
  });

  describe('when user hits enter key', () => {
    it('should submit request', () => {
      const processRequest = jest.spyOn(wrapper.find('CustomQuery').instance(), 'processRequest');
      const preventDefault = jest.fn();
      textField.props().onKeyDown({ keyCode: 13, preventDefault });
      expect(preventDefault).toHaveBeenCalledTimes(1);
      expect(processRequest).toHaveBeenCalledTimes(1);
      expect(submit).toHaveBeenCalledTimes(1);
    });
  });
  describe('when user inputs text', () => {
    it('should change url query string in state', () => {
      const setUrlQs = jest.spyOn(wrapper.find('CustomQuery').props().actions, 'setURLQs');
      textField.props().onChange({ target: { value: 'abc' } });
      expect(wrapper.find('CustomQuery').state('qsParams').textValue).toBe('abc');
      expect(setUrlQs).toHaveBeenCalledTimes(1);
    });
  });
});
