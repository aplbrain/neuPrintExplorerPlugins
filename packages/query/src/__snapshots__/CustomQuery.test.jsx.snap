// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom query Plugin renders correctly 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <div
    className="MuiFormControl-root-1"
    onKeyDown={[Function]}
  >
    <label
      className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
      data-shrink={false}
    >
      Custom Cypher Query
    </label>
    <div
      className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
      onClick={[Function]}
    >
      <div
        className="MuiPrivateTextarea-root-53"
      >
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows="1"
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows={1}
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-invalid={false}
          className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={1}
          style={
            Object {
              "height": 19,
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root-82 MuiButton-root-56 MuiButton-contained-67 MuiButton-containedPrimary-68 MuiButton-raised-70 MuiButton-raisedPrimary-71"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-57"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-85"
    />
  </button>
</div>
`;
